Can't Create Model Sessions on Different GPU

**Describe the bug**
I am on a 2 GPU system, I tried to create one session on each GPU (so 2 sessions in total). I get the following error when it's creating the second session (i.e. creating the first session on GPU 0, and the second session on GPU1)
```
Failed to get allocator for location: OrtAllocatorInfo: [ name:Cuda id:0 mem_type:0 type:1]
```
Changing the creation order (first session on GPU1, second session on GPU0) will gives the same kind of error, but with different id
```
Failed to get allocator for location: OrtAllocatorInfo: [ name:Cuda id:1 mem_type:0 type:1]
```

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 0.4.0
- Python version:
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source): gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
- CUDA/cuDNN version: 10.1.105 / 7.5.0.56
- GPU model and memory: Tesla P100, 16 GB

**To Reproduce**
- Compile the code (modified from c_api_example) below, and run the executable
- the model file "int32_int32_int32.onnx" is just a simple model that takes two inputs, returns two outputs for addition and subtraction. Will provide the generation script if requested, but the model itself seems irrelevant to the issue
```
#include <assert.h>
#include <core/providers/cuda/cuda_provider_factory.h>
#include <core/session/onnxruntime_c_api.h>
#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <vector>

//*****************************************************************************
// helper function to check for status
#define CHECK_STATUS(expr)                               \
  {                                                      \
    OrtStatus* onnx_status = (expr);                     \
    if (onnx_status != NULL) {                           \
      const char* msg = OrtGetErrorMessage(onnx_status); \
      fprintf(stderr, "%s\n", msg);                      \
      OrtReleaseStatus(onnx_status);                     \
      exit(1);                                           \
    }                                                    \
  }

int
main(int argc, char* argv[])
{
  //===================== Env Scope ======================
  // initialize  enviroment...one enviroment per process
  // enviroment maintains thread pools and other state info
  OrtEnv* env = nullptr;
  // [TODO] look into OrtCreateEnvWithCustomLogger()
  CHECK_STATUS(OrtCreateEnv(ORT_LOGGING_LEVEL_WARNING, "test", &env));

  //===================== Model Scope ======================
  // session option also specifies GPU device
  // but other session option can be set "globally", just clone the session
  // when setting actual device
  OrtSessionOptions* session_options = OrtCreateSessionOptions();
  OrtSetSessionThreadPoolSize(session_options, 1);

  // disable graph optimization
  OrtSetSessionGraphOptimizationLevel(session_options, 0);

  //===================== Session Scope ======================
  // create multiple sessions and load models into memory
  std::vector<const char*> model_paths;
  {
    // two instances of the same model
    model_paths.emplace_back();
    model_paths.back() = "int32_int32_int32.onnx";
    model_paths.emplace_back();
    model_paths.back() = "int32_int32_int32.onnx";
  }

  std::vector<OrtSession*> sessions(model_paths.size());
  for (size_t idx = 0; idx < model_paths.size(); idx++) {
    // To deploy on different device, need to set provider on cloned option
    // as you can remove a execution provider once it is appended
    OrtSessionOptions* context_options =
        OrtCloneSessionOptions(session_options);
    
    // Use CUDA, device 0. CPU if not set.
    // Use "idx" as we know we have two models and two devices
    OrtSessionOptionsAppendExecutionProvider_CUDA(context_options, idx);
    std::cout << "here" << std::endl;
    // [TODO] change session options for different session
    CHECK_STATUS(OrtCreateSession(
        env, model_paths[idx], context_options, &sessions[idx]));
    std::cout << "after here" << std::endl;
    // session-wise
    OrtReleaseSessionOptions(context_options);
  }
}
```

**Expected behavior**
Expect to be able to create session on different GPU at the same time. Right now, I can create two sessions on the same GPU (GPU0 only or GPU1 only).

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

We should fix the following warning:
```
/home/chasun/src/onnxruntime/onnxruntime/core/providers/cuda/cuda_allocator.h:24:13: warning: private field 'device_id_' is not used [-Wunused-private-field]
  const int device_id_;
```

--- a/onnxruntime/core/providers/cuda/cuda_allocator.cc
+++ b/onnxruntime/core/providers/cuda/cuda_allocator.cc
@@ -20,7 +20,7 @@ void CUDAAllocator::CheckDevice() const {
   // if it's expected to change, call cudaSetDevice instead of the check
   int current_device;
   CUDA_CALL_THROW(cudaGetDevice(&current_device));
-  ORT_ENFORCE(current_device == device_id_);
+  ORT_ENFORCE(current_device == info_.id);
 #endif
 }
 

--- a/onnxruntime/core/providers/cuda/cuda_allocator.h
+++ b/onnxruntime/core/providers/cuda/cuda_allocator.h
@@ -11,7 +11,7 @@ constexpr const char* CUDA_PINNED = "CudaPinned";
 
 class CUDAAllocator : public IDeviceAllocator {
  public:
-  CUDAAllocator(int device_id) : device_id_(device_id), info_(CUDA, OrtAllocatorType::OrtDeviceAllocator, device_id, OrtMemTypeDefault) {}
+  CUDAAllocator(int device_id) : info_(CUDA, OrtAllocatorType::OrtDeviceAllocator, device_id, OrtMemTypeDefault) {}
   virtual void* Alloc(size_t size) override;
   virtual void Free(void* p) override;
   virtual const OrtAllocatorInfo& Info() const override;
@@ -21,7 +21,6 @@ class CUDAAllocator : public IDeviceAllocator {
   void CheckDevice() const;
 
  private:
-  const int device_id_;
   const OrtAllocatorInfo info_;
 };
 

--- a/onnxruntime/core/providers/cuda/cuda_execution_provider.cc
+++ b/onnxruntime/core/providers/cuda/cuda_execution_provider.cc
@@ -40,13 +40,17 @@ ONNX_OPERATOR_KERNEL_EX(
 
 }  // namespace cuda
 
-thread_local std::shared_ptr<CUDAExecutionProvider::PerThreadContext> CUDAExecutionProvider::per_thread_context_;
-thread_local AllocatorPtr CUDAExecutionProvider::per_thread_default_allocator_;
+thread_local std::unique_ptr<CUDAExecutionProvider::PerThreadContextMap> CUDAExecutionProvider::per_thread_context_map_;
 
 CUDAExecutionProvider::PerThreadContext::PerThreadContext(int device_id) {
   CUDA_CALL_THROW(cudaSetDevice(device_id));
   CUBLAS_CALL_THROW(cublasCreate(&cublas_handle_));
   CUDNN_CALL_THROW(cudnnCreate(&cudnn_handle_));
+
+  DeviceAllocatorRegistrationInfo default_allocator_info(
+      {OrtMemTypeDefault,
+       [](int id) { return std::make_unique<CUDAAllocator>(id); }, std::numeric_limits<size_t>::max()});
+  allocator_ = CreateAllocator(default_allocator_info, device_id);
 }
 
 CUDAExecutionProvider::PerThreadContext::~PerThreadContext() {
@@ -92,16 +96,36 @@ CUDAExecutionProvider::~CUDAExecutionProvider() {
   ReleasePerThreadStuffs();
 }
 
-void CUDAExecutionProvider::ReleasePerThreadStuffs() const {
-  if (per_thread_default_allocator_) {
-    std::lock_guard<OrtMutex> lock(default_allocator_pool_mutex_);
-    default_allocator_pool_.push_back(per_thread_default_allocator_);
-    per_thread_default_allocator_.reset();
+CUDAExecutionProvider::PerThreadContext& CUDAExecutionProvider::GetPerThreadContext() const {
+  if (per_thread_context_map_ == nullptr) {
+    per_thread_context_map_ = std::make_unique<PerThreadContextMap>();
   }
-  if (per_thread_context_) {
+
+  auto* p = per_thread_context_map_.get();
+  if (p->count(this) == 0) {
     std::lock_guard<OrtMutex> lock(context_pool_mutex_);
-    context_pool_.push_back(per_thread_context_);
-    per_thread_context_.reset();
+    if (context_pool_.empty()) {
+      p->insert(std::make_pair(this, std::make_shared<PerThreadContext>(device_id_)));
+    } else {
+      p->insert(std::make_pair(this, context_pool_.back()));
+      context_pool_.pop_back();
+    }
+  }
+  return *(p->at(this));
+}
+
+void CUDAExecutionProvider::ReleasePerThreadStuffs() const {
+  if (per_thread_context_map_ != nullptr && !per_thread_context_map_->empty()) {
+    auto iter_ctx = per_thread_context_map_->find(this);
+    if (iter_ctx != per_thread_context_map_->end()) {
+      std::lock_guard<OrtMutex> lock(context_pool_mutex_);
+      context_pool_.push_back(iter_ctx->second);
+      per_thread_context_map_->erase(iter_ctx);
+      // Release TLS if empty to avoid memory leak report
+      if (per_thread_context_map_->empty()) {
+        per_thread_context_map_.reset(nullptr);
+      }
+    }
   }
 }
 
@@ -110,19 +134,7 @@ AllocatorPtr CUDAExecutionProvider::GetAllocator(int id, OrtMemType mem_type) co
   // A hypothesis is that arena allocator is not aligned with CUDA output cache, and data from different kernel writes may
   // cause cacheline to contain dirty data.
   if (mem_type == OrtMemTypeDefault) {
-    if (!per_thread_default_allocator_) {
-      std::lock_guard<OrtMutex> lock(default_allocator_pool_mutex_);
-      if (default_allocator_pool_.empty()) {
-        DeviceAllocatorRegistrationInfo default_allocator_info(
-            {OrtMemTypeDefault,
-             [](int id) { return std::make_unique<CUDAAllocator>(id); }, std::numeric_limits<size_t>::max()});
-        per_thread_default_allocator_ = CreateAllocator(default_allocator_info, device_id_);
-      } else {
-        per_thread_default_allocator_ = default_allocator_pool_.back();
-        default_allocator_pool_.pop_back();
-      }
-    }
-    return per_thread_default_allocator_;
+    return GetPerThreadContext().GetAllocator();
   } else {
     return IExecutionProvider::GetAllocator(id, mem_type);
   }
@@ -137,7 +149,7 @@ void CUDAExecutionProvider::AddDeferredReleaseCPUPtr(void* p) {
   // when not running in InferenceSession (e.g. Test)
   // it's OK to not remember the deferred release ptr
   // as the actual memory will be cleaned in arena allocator dtor
-  auto current_deferred_release_event = per_thread_context_->GetCurrentDeferredReleaseEvent();
+  auto current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   if (current_deferred_release_event) {
     std::lock_guard<OrtMutex> lock(deferred_release_cpu_ptr_mutex_);
     auto iter = deferred_release_cpu_ptr_.find(current_deferred_release_event);
@@ -167,26 +179,16 @@ Status CUDAExecutionProvider::OnRunStart() {
       ++it;
     }
   }
-  // start a new per_thread context, store in TLS
-  {
-    std::lock_guard<OrtMutex> ctx_lock(context_pool_mutex_);
-    if (context_pool_.empty()) {
-      per_thread_context_ = std::make_shared<PerThreadContext>(device_id_);
-    } else {
-      per_thread_context_ = context_pool_.back();
-      context_pool_.pop_back();
-    }
-  }
-  auto& current_deferred_release_event = per_thread_context_->GetCurrentDeferredReleaseEvent();
+
+  auto& current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventCreate(&current_deferred_release_event, cudaEventDisableTiming));
   deferred_release_cpu_ptr_.emplace(current_deferred_release_event, DeferredReleaseCPUPtrs());
   return Status::OK();
 }
 
 Status CUDAExecutionProvider::OnRunEnd() {
-  ORT_RETURN_IF_NOT(per_thread_context_ != nullptr);
   // record deferred release event on default stream, and release per_thread_context
-  auto current_deferred_release_event = per_thread_context_->GetCurrentDeferredReleaseEvent();
+  auto current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventRecord(current_deferred_release_event, nullptr));
   ReleasePerThreadStuffs();
   std::lock_guard<OrtMutex> lock(deferred_release_cpu_ptr_mutex_);

--- a/onnxruntime/core/providers/cuda/cuda_execution_provider.h
+++ b/onnxruntime/core/providers/cuda/cuda_execution_provider.h
@@ -43,18 +43,11 @@ class CUDAExecutionProvider : public IExecutionProvider {
   Status CopyTensor(const Tensor& src, Tensor& dst, int exec_queue_id) const override;
 
   cublasHandle_t PerThreadCublasHandle() {
-    // Assure each thread has its TLS context.
-    if (!per_thread_context_)
-      per_thread_context_ = std::make_shared<PerThreadContext>(device_id_);
-    return per_thread_context_->CublasHandle();
+    return GetPerThreadContext().CublasHandle();
   }
 
   cudnnHandle_t PerThreadCudnnHandle() {
-    // Assure each thread has its TLS context.
-    // TODO: improve its performance when calling cuda functions from multiple threads.
-    if (!per_thread_context_)
-      per_thread_context_ = std::make_shared<PerThreadContext>(device_id_);
-    return per_thread_context_->CudnnHandle();
+    return GetPerThreadContext().CudnnHandle();
   }
 
   cudaStream_t GetStream(int queue_id) const {
@@ -64,10 +57,7 @@ class CUDAExecutionProvider : public IExecutionProvider {
 
   template <typename T>
   const T* GetConstOnes(size_t count) {
-    // Assure each thread has its TLS context.
-    if (!per_thread_context_)
-      per_thread_context_ = std::make_shared<PerThreadContext>(device_id_);
-    return per_thread_context_->template GetConstOnes<T>(count);
+    return GetPerThreadContext().template GetConstOnes<T>(count);
   }
 
   void AddDeferredReleaseCPUPtr(void* p);
@@ -136,6 +126,10 @@ class CUDAExecutionProvider : public IExecutionProvider {
       }
     }
 
+    AllocatorPtr GetAllocator() const {
+      return allocator_;
+    }
+
    private:
     cublasHandle_t cublas_handle_ = nullptr;
     cudnnHandle_t cudnn_handle_ = nullptr;
@@ -148,22 +142,19 @@ class CUDAExecutionProvider : public IExecutionProvider {
     std::unique_ptr<cuda::IConstantBuffer<float>> constant_ones_float_;
     std::unique_ptr<cuda::IConstantBuffer<double>> constant_ones_double_;
     std::unique_ptr<cuda::IConstantBuffer<half>> constant_ones_half_;
+
+    AllocatorPtr allocator_;
   };
 
   // thread local context during execution
-  static thread_local std::shared_ptr<PerThreadContext> per_thread_context_;
-
-  // thread local GPU memory allocator. could be used before execution
-  static thread_local AllocatorPtr per_thread_default_allocator_;
-
-  // reuse thread local GPU memory allocator for memory pattern
-  mutable std::deque<AllocatorPtr> default_allocator_pool_;
-  mutable OrtMutex default_allocator_pool_mutex_;
+  using PerThreadContextMap = std::unordered_map<const CUDAExecutionProvider*, std::shared_ptr<PerThreadContext>>;
+  static thread_local std::unique_ptr<PerThreadContextMap> per_thread_context_map_;
 
   // reuse thread local context
   mutable std::deque<std::shared_ptr<PerThreadContext>> context_pool_;
   mutable OrtMutex context_pool_mutex_;
 
+  PerThreadContext& GetPerThreadContext() const;
   void ReleasePerThreadStuffs() const;
 };
 

